### PR TITLE
[Kernel] Change string comparing from UTF16 to UTF8

### PR DESCRIPTION
**Pull request by ilicmarkodb**
## Description
Changed string comparing from UTF16 to UTF8. This fixes comparison issues around the characters with surrogate pairs.

## How was this patch tested?
Tests added to `DefaultExpressionEvaluatorSuite.scala`

### CODE REVIEW

File: kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/DeltaTableReadsSuite.scala [hunk 1]
```diff
@@ -17,19 +17,24 @@ package io.delta.kernel.defaults
 
 import io.delta.golden.GoldenTableUtils.goldenTablePath
 import io.delta.kernel.exceptions.{InvalidTableException, KernelException, TableNotFoundException}
+import io.delta.kernel.expressions.{Literal, ScalarExpression}
```
Comment (vkorukanti) [inline kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/DeltaTableReadsSuite.scala:20]: unrelated changes in this file?

File: kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/internal/expressions/DefaultExpressionEvaluatorSuite.scala [hunk 2]
```diff
@@ -607,6 +607,44 @@ class DefaultExpressionEvaluatorSuite extends AnyFunSuite with ExpressionSuiteBa
       ),
       (ofDate(-12123), ofDate(123123), ofDate(-12123), ofNull(DateType.DATE)),
       (ofString("apples"), ofString("oranges"), ofString("apples"), ofNull(StringType.STRING)),
+      (
+        ofString("apples"),
+        ofString("oranges"),
+        ofString("apples"),
+        ofNull(StringType.STRING)
+      ),
```
Comment (vkorukanti) [inline kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/internal/expressions/DefaultExpressionEvaluatorSuite.scala:615]: same test case as the one above?

File: kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/internal/expressions/DefaultExpressionEvaluatorSuite.scala [hunk 3]
```diff
@@ -607,6 +607,44 @@ class DefaultExpressionEvaluatorSuite extends AnyFunSuite with ExpressionSuiteBa
...
+      (
+        ofString("abc"),
+        ofString("abcd"),
+        ofString("abc"),
+        ofNull(StringType.STRING)
+      ),
```
Comment (vkorukanti) [inline kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/internal/expressions/DefaultExpressionEvaluatorSuite.scala:621]:
```suggestion
      (ofString("abc"), ofString("abcd"), ofString("abc"), ofNull(STRING)),
```

Comment (vkorukanti) [reply kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/internal/expressions/DefaultExpressionEvaluatorSuite.scala:621]: same for other test cases.

File: kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/internal/expressions/DefaultExpressionEvaluatorSuite.scala [hunk 4]
```diff
@@ -607,6 +607,44 @@ class DefaultExpressionEvaluatorSuite extends AnyFunSuite with ExpressionSuiteBa
...
+        ofString("\u00EE"),
+        ofString("\00BB"),
+        ofNull(StringType.STRING)
+      ),
+      (
+        ofString("\uFFFD"),
+        ofString("\uD83C\uDF3C"),
+        ofString("\uFFFD"),
+        ofNull(StringType.STRING)
+        // scalastyle:on nonascii
```
Comment (vkorukanti) [inline kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/internal/expressions/DefaultExpressionEvaluatorSuite.scala:646]: are there more tests we could copy for Spark or Delta-Spark around other surrogate pairs and unsigned char comparison?

**Review by vkorukanti — COMMENTED** [review]
Thanks for catching this and fixing it.

**Review by stefankandic — APPROVED** [review]
